### ISSUE TITLE
jackson-databind 2.11.0-SNAPSHOT no longer supports Scala Either[L,R]

### ISSUE THREAD

**Issue by pjfanning** [FasterXML/jackson-databind#2569]
I've isolated the change in jackson-databind to changes of 9 Nov 2019.

* https://travis-ci.org/FasterXML/jackson-module-scala/jobs/621987475
* https://github.com/FasterXML/jackson-module-scala/pull/439

**Comment by cowtowncoder** [comment]
From warnings, this seems relevant:

```
[warn]         ctxt.reportWrongTokenException(jp, JsonToken.END_ARRAY,
[warn]              ^
[warn] /home/travis/build/FasterXML/jackson-module-scala/src/main/scala/com/fasterxml/jackson/module/scala/ser/EitherSerializer.scala:60: method findConvertingContentSerializer in class StdSerializer is deprecated: see corresponding Javadoc for more information.
[warn]     ser = Option(findConvertingContentSerializer(prov, prop, ser.orNull))
[warn]                  ^
[warn] 5 warnings found
```

and as per Javadocs

```
    /**
     * @deprecated Since 2.9 use {link {@link #findContextualConvertingSerializer} instead
     */
    @Deprecated
    protected JsonSerializer<?> findConvertingContentSerializer(SerializerProvider provider,
            BeanProperty prop, JsonSerializer<?> existingSerializer)
```

Not sure if what might have changed since deprecation was bit earlier but typically behavior of deprecated methods is less well tested after deprecated (that is, test code is updated to use non-deprecated functionality, most of the time).

**Comment by pjfanning** [comment]
I removed the use of the findConvertingContentSerializer method but the same issue occurs.

**Comment by pjfanning** [comment]
In my testing, I isolated the change that caused the issue to the commits of Nov 9. https://github.com/FasterXML/jackson-databind/commits/2.11?after=2f89b9ca71f8e009f50b86506dc5f11804b5ad94+34

**Comment by cowtowncoder** [comment]
Did not realize that `Either` was a `ReferenceType`. This could be related since registration of serializers is based on subtype of `JavaType`.

Unfortunately I can't quite decipher Scala code with traits and all so I don't know where and how `EitherSerializer` is registered -- but I suspect it may be relying on older mechanism wherein default ("bean" / POJO) lookup was (also) called, not just one added for `ReferenceType` (which itself was added in 2.6).

**Comment by pjfanning** [comment]
@cowtowncoder I fixed jackson-module-scala with https://github.com/FasterXML/jackson-module-scala/commit/27186457241514b0762032a02270e8e6ac3b9184 - in the past, findBeanDeserializer worked but now findReferenceDeserializer is needed

**Comment by cowtowncoder** [comment]
@pjfanning excellent, thank you.

I am not sure how this could be done, but it would be great to figure out a way to do Travis builds to help figure out breaking changes in `jackson-databind` (or theoretically `jackson-core`) wrt downstream dependencies, during development of new minor versions.
This would make it much easier to resolve issues if they were discovered earlier.

**Comment by pjfanning** [comment]
@cowtowncoder I set up a daily build on jackson-module-scala 2.11 branch - based on https://docs.travis-ci.com/user/cron-jobs/ - I'll see if it runs as expected

**Comment by cowtowncoder** [comment]
the only (?) trick really would be to trigger that build when jackson-databind or jackson-core runs. Perhaps it could be triggered by availability of a new snapshot version at Sonatype snapshot repo?

**Comment by pjfanning** [comment]
the cron-based build I set up for jackson-module-scala pulls in latest jackson 2.11.0-SNAPSHOT jars

**Comment by cowtowncoder** [comment]
@pjfanning so it runs on regular schedule, in addition to being triggered by new commits? That would work.

**Comment by pjfanning** [comment]
yes it builds off both commits and a daily schedule

**Comment by cowtowncoder** [comment]
Not related to this issue which is now fixed it sounds like (yay!), but completing #2527 which supports Kotlin's approach via new `AnnotationIntrospector` method. Not sure it would be useful for Scala, but will mention for sake of completeness.
